services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/TemplateHelper.java

```diff
@@ -0,0 +1,145 @@
+         * <p>
+         * If the RDNs list contains multiple occurrences of the same attribute, the value
+         * of the first occurrence is returned.
+         *
+         * @param rdns The list of RDNs.
+         * @return The attribute value or {@code null} if the attribute is not available in the RDNs list.
+         * @throws NullPointerException if rdns is {@code null}.
+         */
+        String getValue(final List<Rdn> rdns) {
+            Objects.requireNonNull(rdns, "rdns list must not be null");
+
+            return rdns
+                    .stream()
+                    .filter(rdn -> this.toString().equalsIgnoreCase(rdn.getType()))
+                    .findFirst()
+                    .map(Rdn::getValue)
+                    .filter(String.class::isInstance)
+                    .map(String.class::cast)
+                    .orElse(null);
+        }
+    }
+
+    /**
+     * Fills placeholders in template with corresponding attribute values from the given subject DN.
+     * <p>
+     * The following placeholders are supported.
+     * <ul>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_DN} for <em>Subject Distinguished Name (DN)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_CN} for <em>Common Name (CN)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_OU} for <em>Organizational Unit Name (OU)</em></li>
+     * <li>{@value RegistryManagementConstants#PLACEHOLDER_SUBJECT_O} for <em>Organization Name (O)</em></li>
+     * </ul>
+     *
+     * @param template The template to be filled.
+     * @param subjectDN The subject DN.
+     * @return The filled template.
+     * @throws IllegalArgumentException if the subject DN is not valid or any of the attributes
+     *                                  configured in the template are not present in the subject DN.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public static String fill(final String template, final String subjectDN) {
+        Objects.requireNonNull(template, "template must not be null");
+        Objects.requireNonNull(subjectDN, "subjectDN must not be null");
+
+        try {
+            final List<Rdn> rdns = new LdapName(subjectDN).getRdns();
+            String result = template.replaceAll(QUOTED_PLACEHOLDER_SUBJECT_DN, subjectDN);
+            result = fillAttribute(Attribute.CN, result, rdns);
+            result = fillAttribute(Attribute.OU, result, rdns);
+            return fillAttribute(Attribute.O, result, rdns);
+        } catch (InvalidNameException e) {
+            throw new IllegalArgumentException(String.format("subject DN [%s] is not valid", subjectDN));
+        }
+    }
+
+    private static String fillAttribute(final Attribute attribute, final String template,
+            final List<Rdn> rdns) {
+        if (template.contains(attribute.getPlaceHolder())) {
```

Comment:
FMPOV it doesn't make much sense to first scan the template string for occurrence of the placeholder as that is what the String.replaceAll method already does as well, doesn't it?

Reply:
This check is there to handle the scenario when placeholder for an attribute is configured in the template but there is no corresponding attribute/value provided in the client certificate. In this case an `IllegalArgumentException` is thrown with an appropriate error reason.